File: book_catalog.js
// book_catalog.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');
const { program } = require('commander');
const bwipjs = require('bwip-js');

const DATA_FILE = 'books.json';
const BARCODE_DIR = 'barcodes';

class Book {
    constructor(title, author, isbn, year, publisher) {
        this.title = title;
        this.author = author;
        this.isbn = isbn;
        this.year = year;
        this.publisher = publisher;
    }
}

class Catalog {
    constructor() {
        this.books = [];
        this.load();
    }

    load() {
        if (fs.existsSync(DATA_FILE)) {
            const data = JSON.parse(fs.readFileSync(DATA_FILE));
            this.books = data.map(b => new Book(b.title, b.author, b.isbn, b.year, b.publisher));
        }
    }

    save() {
        fs.writeFileSync(DATA_FILE, JSON.stringify(this.books, null, 2));
    }

    add(title, author, isbn, year, publisher) {
        for (const b of this.books) {
            if (b.isbn === isbn) {
                console.log(`Book with ISBN ${isbn} already exists.`);
                return;
            }
        }
        const book = new Book(title, author, isbn, year, publisher);
        this.books.push(book);
        this.save();
        console.log(`✅ Added: "${title}" by ${author} (ISBN: ${isbn})`);
    }

    list() {
        if (this.books.length === 0) {
            console.log('No books in catalog.');
            return;
        }
        console.log('\n📋 All Books:');
        this.books.forEach((b, i) => {
            console.log(`${i+1}. ${b.title} (${b.isbn}) – ${b.author} (${b.year})`);
        });
    }

    search(term) {
        const lower = term.toLowerCase();
        const results = this.books.filter(b =>
            b.title.toLowerCase().includes(lower) ||
            b.author.toLowerCase().includes(lower) ||
            b.isbn.includes(term)
        );
        if (results.length === 0) {
            console.log('No matching books.');
            return;
        }
        console.log(`\n🔍 Search results for "${term}":`);
        results.forEach(b => {
            console.log(`${b.title} – ${b.author} (${b.isbn})`);
        });
    }

    show(isbn) {
        const book = this.books.find(b => b.isbn === isbn);
        if (!book) {
            console.log(`Book with ISBN ${isbn} not found.`);
            return;
        }
        console.log(`\n📖 Details for ${isbn}:`);
        console.log(`Title: ${book.title}`);
        console.log(`Author: ${book.author}`);
        console.log(`ISBN: ${book.isbn}`);
        console.log(`Year: ${book.year}`);
        console.log(`Publisher: ${book.publisher}`);
        const barcodePath = path.join(BARCODE_DIR, `${isbn}.png`);
        if (fs.existsSync(barcodePath)) {
            console.log(`Barcode: ${barcodePath}`);
        } else {
            console.log('Barcode not generated yet.');
        }
    }

    async generateBarcode(isbn) {
        if (!/^\d{13}$/.test(isbn)) {
            console.log('Invalid ISBN. Must be 13 digits.');
            return;
        }
        const book = this.books.find(b => b.isbn === isbn);
        if (!book) {
            console.log(`Book with ISBN ${isbn} not found.`);
            return;
        }
        try {
            fs.mkdirSync(BARCODE_DIR, { recursive: true });
            const png = await bwipjs.toBuffer({
                bcid: 'ean13',
                text: isbn,
                scale: 3,
                height: 10,
                includetext: true,
            });
            const filepath = path.join(BARCODE_DIR, `${isbn}.png`);
            fs.writeFileSync(filepath, png);
            console.log(`✅ Barcode generated: ${filepath}`);
        } catch (err) {
            console.error('Error generating barcode:', err);
        }
    }
}

program
    .command('add <title> <author> <isbn> <year> <publisher>')
    .action((title, author, isbn, year, publisher) => {
        const catalog = new Catalog();
        catalog.add(title, author, isbn, parseInt(year), publisher);
    });

program
    .command('list')
    .action(() => {
        const catalog = new Catalog();
        catalog.list();
    });

program
    .command('search <term>')
    .action((term) => {
        const catalog = new Catalog();
        catalog.search(term);
    });

program
    .command('show <isbn>')
    .action((isbn) => {
        const catalog = new Catalog();
        catalog.show(isbn);
    });

program
    .command('barcode <isbn>')
    .action(async (isbn) => {
        const catalog = new Catalog();
        await catalog.generateBarcode(isbn);
    });

program.parse(process.argv);

File: book_catalog.rb
# book_catalog.rb
#!/usr/bin/env ruby
require 'json'
require 'optparse'
require 'barby'
require 'barby/outputter/png_outputter'
require 'fileutils'

DATA_FILE = 'books.json'
BARCODE_DIR = 'barcodes'

class Book
  attr_accessor :title, :author, :isbn, :year, :publisher

  def initialize(title, author, isbn, year, publisher)
    @title = title
    @author = author
    @isbn = isbn
    @year = year
    @publisher = publisher
  end

  def to_hash
    { title: @title, author: @author, isbn: @isbn, year: @year, publisher: @publisher }
  end

  def self.from_hash(h)
    new(h['title'], h['author'], h['isbn'], h['year'], h['publisher'])
  end
end

class Catalog
  attr_reader :books

  def initialize
    @books = []
    load
  end

  def load
    if File.exist?(DATA_FILE)
      data = JSON.parse(File.read(DATA_FILE))
      @books = data.map { |h| Book.from_hash(h) }
    end
  end

  def save
    File.write(DATA_FILE, JSON.pretty_generate(@books.map(&:to_hash)))
  end

  def add(title, author, isbn, year, publisher)
    if @books.any? { |b| b.isbn == isbn }
      puts "Book with ISBN #{isbn} already exists."
      return
    end
    book = Book.new(title, author, isbn, year, publisher)
    @books << book
    save
    puts "✅ Added: \"#{title}\" by #{author} (ISBN: #{isbn})"
  end

  def list
    if @books.empty?
      puts "No books in catalog."
      return
    end
    puts "\n📋 All Books:"
    @books.each_with_index do |b, i|
      puts "#{i+1}. #{b.title} (#{b.isbn}) – #{b.author} (#{b.year})"
    end
  end

  def search(term)
    lower = term.downcase
    results = @books.select do |b|
      b.title.downcase.include?(lower) ||
      b.author.downcase.include?(lower) ||
      b.isbn.include?(term)
    end
    if results.empty?
      puts "No matching books."
      return
    end
    puts "\n🔍 Search results for \"#{term}\":"
    results.each { |b| puts "#{b.title} – #{b.author} (#{b.isbn})" }
  end

  def show(isbn)
    book = @books.find { |b| b.isbn == isbn }
    unless book
      puts "Book with ISBN #{isbn} not found."
      return
    end
    puts "\n📖 Details for #{isbn}:"
    puts "Title: #{book.title}"
    puts "Author: #{book.author}"
    puts "ISBN: #{book.isbn}"
    puts "Year: #{book.year}"
    puts "Publisher: #{book.publisher}"
    barcode_path = File.join(BARCODE_DIR, "#{isbn}.png")
    if File.exist?(barcode_path)
      puts "Barcode: #{barcode_path}"
    else
      puts "Barcode not generated yet."
    end
  end

  def generate_barcode(isbn)
    unless isbn.match?(/^\d{13}$/)
      puts "Invalid ISBN. Must be 13 digits."
      return
    end
    book = @books.find { |b| b.isbn == isbn }
    unless book
      puts "Book with ISBN #{isbn} not found."
      return
    end
    require 'barby/barcode/ean_13'
    FileUtils.mkdir_p(BARCODE_DIR)
    begin
      barcode = Barby::EAN13.new(isbn)
      outputter = Barby::PngOutputter.new(barcode)
      outputter.to_png(height: 60, margin: 10, xdim: 2)
      File.open(File.join(BARCODE_DIR, "#{isbn}.png"), 'wb') do |f|
        f.write(outputter.to_png(height: 60, margin: 10, xdim: 2))
      end
      puts "✅ Barcode generated: #{BARCODE_DIR}/#{isbn}.png"
    rescue => e
      puts "Error generating barcode: #{e.message}"
    end
  end
end

options = {}
$command = ARGV.shift
if $command.nil?
  puts "Usage: book_catalog.rb <command> [options]"
  exit 1
end

catalog = Catalog.new

case $command
when 'add'
  if ARGV.size < 5
    puts "add <title> <author> <isbn> <year> <publisher>"
    exit 1
  end
  title, author, isbn, year, publisher = ARGV[0], ARGV[1], ARGV[2], ARGV[3].to_i, ARGV[4]
  catalog.add(title, author, isbn, year, publisher)
when 'list'
  catalog.list
when 'search'
  term = ARGV.shift
  catalog.search(term) if term
when 'show'
  isbn = ARGV.shift
  catalog.show(isbn) if isbn
when 'barcode'
  isbn = ARGV.shift
  catalog.generate_barcode(isbn) if isbn
else
  puts "Unknown command"
end
